### DISCUSSION TITLE
Struggling with DAP and code lenses

### DISCUSSION THREAD

**Discussion by susliko** [scalameta/nvim-metals#319] [Q&A] [answered]
I'm  trying to get working "Using code lenses" part of the [DAP integration fetature](https://github.com/scalameta/nvim-metals/discussions/279#discussioncomment-1774771)
What I did:
1. [Configured dap initialization](https://github.com/susliko/dotfiles/blob/master/nvim/.config/nvim/lua/susliko/lang/scala.lua#L38) on attach to a buffer
2. Created new scala project from the `hello-world.g8` template
3. Invoked `:lua vim.lsp.codelens.run()` on the line with Main and chose `run`
4. Got an error 
```
Language server does not support command `metals-run-session-start`. 
This command may require a client extension
```
![image](https://user-images.githubusercontent.com/15569000/148226948-6160fe04-25d0-4d80-b4e8-3b701c7fd67e.png)

Am I missing some part of the configuration?

**Comment by ckipp01** [comment, marked as answer]
Odd, that message is coming straight from Neovim indicating it's not picking up the commands being set by `nvim-metals`. Can you show me the output of `:lua print(vim.inspect(vim.lsp.buf_get_clients()[1].config.commands))` when you're in that file?

**Reply by susliko** [reply]
It shows
```
{   
  ["metals-debug-session-start"] = <function 1>,       
  ["metals-run-session-start"] = <function 2>     
} 
```

**Reply by ckipp01** [reply]
![image](https://user-images.githubusercontent.com/13974112/148230605-39a13e5f-2b57-47d1-8e16-f5216f8f1cff.png)

Huh, so that command tells me that the `metals-run-session-start` function is there, but when you trigger it via code lens it can't be found. I'm assuming you have the same issue when you choose debug?

**Reply by susliko** [reply]
Yes, debug also results in error:
![image](https://user-images.githubusercontent.com/15569000/148232554-b914f4a0-6e17-4251-b26f-4b1d48358111.png)

**Reply by ckipp01** [reply]
What commit of `nvim-metals` are you using? I assume a pretty new one because you have the new info format. Also what version of Neovim are you using?

**Reply by susliko** [reply]
Idk if it's related, but `runType = "runOrTestFile"` results in error `LSP[metals] Received invalid runType: runOrTestFile.`

Meanhile `runType = "run"` runs as expected

**Reply by ckipp01** [reply]
Yea that's sort of expected since `runOrTestFile` isn't in the latest stable. Actually can you try that real quick? Try the latest snaphsot of Metals `0.10.9+285-c82cb7d7-SNAPSHOT`? I merged in the `runOrTestFile` maybe a bit too early. Either way, I don't think the snapshot should make a difference since this is Neovim complaining, not Metals.

**Reply by susliko** [reply]
Neovim:
```
nvim --version
  NVIM v0.6.0-dev
  Build type: RelWithDebInfo
  LuaJIT 2.1.0-beta3
```

nvim-metals:
```
 susliko  ~/.local/share/nvim/site/pack/packer/start/nvim-metals   main  git log | head -n 20
commit 2e1257aec1f93759fb2ca7ffabe6a2e4f2034577
Merge: de820ee f967b75
Author: Chris Kipp <ckipp@pm.me>
Date:   Mon Jan 3 15:39:43 2022 +0100

    Merge pull request #317 from ckipp01/info

    refactor: make info window be markdown

commit f967b75f5d52d4c4829dae11b62bc3385d28d898
Author: Chris Kipp <ckipp@pm.me>
Date:   Mon Jan 3 15:32:01 2022 +0100

    refactor: make info window be markdown

commit de820eeffbf40a663cc684df8bad4a559a6361f1
Merge: c13aace 539878c
Author: Chris Kipp <ckipp@pm.me>
Date:   Mon Jan 3 14:14:40 2022 +0100
```

**Reply by ckipp01** [reply]
Also could you update to the latest stable Neovim, v0.6.1?

**Reply by susliko** [reply]
Indeed, the source of error was my old neovim installation
Both run and debug lenses work now 🎉 

Thank you so much for support!

**Reply by ckipp01** [reply]
Nice! Glad it works!

**Comment by V4G4X** [comment]
Can someone help me with as to how I can get my code to run using codelens?  
I can:
1. Run `lua vim.lsp.codelens.refresh()`
2. See the codelens options show up at my Runner: "run | debug"
3. Run `lua vim.lsp.codelens.run`
4. Get options of pressing 1 for run, and 2 for debug.
5. Press 1. 
6. If says "LSP[metals][LSP] Starting debug server!"
7. But no new buffer opens up with "Hello World" 
8. Once I do `lua require("dap").repl.toggle()`, a buffer opens up but it does not take in input. Instead returns: "Evalutation failed because thread is not suspended"

How does this work?

**Reply by ckipp01** [reply]
There actually isn't a way to get input when using the code lens. Instead you'd want to rely on having a configuration that takes input. For example you can have something like this:

```lua
    local dap = require("dap")

    dap.configurations.scala = {
      {
        type = "scala",
        request = "launch",
        name = "Run or test with input",
        metals = {
          runType = "runOrTestFile",
          args = function()
            local args_string = vim.fn.input("Arguments: ")
            return vim.split(args_string, " +")
          end,
        },
      }
    }
```

Then in the file that has the code lens you'd trigger a mapping like:

```lua
vim.keymap.set("n", "<leader>dc", require("dap").continue)
```
When you do this in a Scala file it will look for the scala configurations that exist and should find the one above which will first prompt you for input, and then shoot it over to Metals.
